### ISSUE TITLE
Graph traversal optimization by excluding graph path (stop conditions)

### ISSUE THREAD

**Issue by vladiksun** [apache/age#993]
ArrangoDB allows to optimize the graph traversal by applying a special clause - PRUNE forcing the engine not to go deeper (stop conditions)
This approach drastically improves the speed of a FILTER sections.
For example:

```aql
FOR any_vertex, any_edge, path IN 1..100 OUTBOUND 'account/0878730b-316e-4afb-a45e-6eca52c95bac' edges
        PRUNE any_edge.relation IN ['managed']
        FILTER any_edge.relation NOT IN ['managed']
        RETURN  any_vertex

```

Can this be somehow simulated in Apache Age + Cypher ?

Following the documentation [Exists(Path)](https://age.apache.org/age-manual/master/functions/predicate_functions.html) 
I assumed something like this could be an alternative

```pgsql
select
    properties
from ag_catalog.cypher('test_graph', $$
    MATCH (from:`account` { id: 'cd94fd96-adeb-40a6-bc17-a2f9b51f6296'} )-[any_edge:`edges`*1..100]->(to)
    WITH from, to, last(any_edge) AS edge_out
    WHERE exists( (from)-[ { relation: "managed" } ]-(to) )
    AND NOT edge_out.relation IN ['managed']

    RETURN properties(to)
$$
         ) as (properties ag_catalog.agtype );

```

But got an error \
[42P01] ERROR: missing FROM-clause entry for table "from"

**Comment by humzakt** [comment]
The error message "missing FROM-clause entry for table "from"" is indicating that there's a problem with the alias from in your query. The word from is a reserved keyword in SQL and it's also used in the MATCH clause of the Cypher language. This might be confusing the query parser.

Try changing the alias from to something else that's not a reserved keyword, for example, start_node:


```
select
    properties
from ag_catalog.cypher('test_graph', $$
    MATCH (start_node:`account` { id: 'cd94fd96-adeb-40a6-bc17-a2f9b51f6296'} )-[any_edge:`edges`*1..100]->(to)
    WITH start_node, to, last(any_edge) AS edge_out
    WHERE exists( (start_node)-[ { relation: "managed" } ]-(to) )
    AND NOT edge_out.relation IN ['managed']

    RETURN properties(to)
$$
         ) as (properties ag_catalog.agtype );

```

**Comment by vladiksun** [comment]
@humzakt thanks for taking a look at it.
I tried the way you suggested but still getting the same error.
![image](https://github.com/apache/age/assets/5010724/c58158d9-9c6a-4d20-8955-0e15e4c23b3e)

Probably I am doing this wrong? Is the syntax valid ? Does EXISTS allows to apply a stop condition ?
Can you suggest if it is even possible with Apache AGE ? I am using version 1.3.0

**Comment by TalhaMunir-JD** [comment]
I think age 1.3.0 doesnot support exist function in where clause of cypher query. It was not implemented in 1.3.0

**Comment by waleedahmed0001** [comment]
The syntax of exist clause is mentioned below:

**Syntax:**
exists(property)

Returns:
An agtype boolean

Arguments:
property | A property from a vertex or edge

A query using the exist is mentioned below:

```
SELECT *
FROM cypher('graph_name', $$
     MATCH (n)
     WHERE exists(n.surname)
     RETURN n.first_name, n.last_name
$$) as (first_name agtype, last_name agtype);
```

**Exists(Path)**
EXISTS(path) returns true if for the given path, there already exists the given path.

```
SELECT *
FROM cypher('graph_name', $$
     MATCH (n)
     WHERE exists((n)-[]-({name: 'Willem Defoe'}))
     RETURN n.full_name
$$) as (full_name agtype);
```

You can get more help from this [documentation](https://age.apache.org/age-manual/master/functions/predicate_functions.html#exists-path)
